src/nn/residual.rs

```diff
@@ -1,52 +1,105 @@
 use crate::prelude::*;
 
-/// A residual connection around `F`: `F(x) + x`,
+/// A residual connection `R` around `F`: `F(x) + R(x)`,
 /// as introduced in [Deep Residual Learning for Image Recognition](https://arxiv.org/abs/1512.03385).
 ///
 /// # Generics
 /// - `F`: The underlying module to do a skip connection around.
+/// - `R`: The underlying residual module
 ///
 /// # Examples
 /// ```rust
 /// # use dfdx::prelude::*;
-/// let module: Residual<ReLU> = Default::default();
+/// let module: Residual<ReLU, Square> = Default::default();
 /// let x = Tensor1D::new([-2.0, -1.0, 0.0, 1.0, 2.0]);
 /// let y = module.forward(x);
-/// assert_eq!(y.data(), &[-2.0, -1.0, 0.0, 2.0, 4.0]);
+/// assert_eq!(y.data(), &[4.0, 1.0, 0.0, 2.0, 6.0]);
 /// ```
 #[derive(Debug, Clone, Default)]
-pub struct Residual<F>(F);
+pub struct Residual<F, R>(F, R);
```

Comment:
I like the idea of having both Residual & ResidualAdd. I imagine most people will probably use ResidualAdd, so let's rename them
- `Residual` does `f(x) + x`
- `GeneralizedResidual` does `f(x) + r(x)`

Reply:
What about `SkipConnection` instead of `GeneralizedResidual`?

Reply:
ok, after re-reading some wiki pages i think that i understand it better now - but shouldn't the gradients of `F` and `R` have to be summed AFTER the backprop of `F` and `R` (so the addition would have to be pushed to the tape first?) or am I not understanding `binary_map` ?

Reply:
All the operations on the tape are executed in reverse, so for this function the add/binary_map backward op will be executed, then the f(x) backward op is executed using the gradient from add(), and then the r(x) backward ops using gradient from add(). Does that help?

Let's keep both Residual and SkipConnection, and then GeneralizedResidual. SkipConnection can just be a type alias (`type SkipConnection<F, R> = Residual<F, R>;`

Reply:
do you have an idea why the numpy test fails?

Reply:
> All the operations on the tape are executed in reverse, so for this function the add/binary_map backward op will be executed, then the f(x) backward op is executed using the gradient from add(), and then the r(x) backward ops using gradient from add(). Does that help?

but isn't the gradient to be used for f(x) and r(x) the same? and shouldn't the layer l before f and r get df + dr as the gradient?

Reply:
> Let's keep both Residual and SkipConnection, and then GeneralizedResidual. SkipConnection can just be a type alias (`type SkipConnection<F, R> = Residual<F, R>;`

Are you sure about creating a type alias? Then it would probably make sense to add an alias `Dense` for `Linear` ...

Reply:
> Does that help?

but what should be added for the gradients? and what about the layer before f and r? shouldn't the gradient there be the sum of df and dr?

Reply:
The gradients are handled automatically by calling add (assuming F & R properly add their derivatives & backward ops to the tape). That's the nice thing about all the stuff in nn - none of it needs to know anything about gradients. All of the functions in tensor_ops handle gradient related things, nn just pieces them together.

This is the key piece about auto diff - you can implement gradient stuff for the lower level building blocks (like add, mul, apply some function, etc), and then piece them together automatically without having to explicitly handle the gradients.

Does that help?

Reply:
i know that, but wouldn't the layer in the forward pass before the residual connection require the addition of the gradients of the residual connection? so the gradient for the layer before the residual connection needs a backward_op pushed _before_ the backward_ops are pushed for the residual layer? how is this done in add?

Reply:
Yeah so the layer before has a gradient of `derivative * result_grad`, where result_grad is calculated by this residual layer during the backward pass (and is the gradient of the input tensor in the forward method). That's basically handled by each tensor having an id associated with it, and the id is used to retrieve the gradient during the backward pass. So the result tensor from the layer before has the same id as the input into the residual layer.

Reply:
> where result_grad is calculated by this residual layer during the backward pass (and is the gradient of the input tensor in the forward method).

so main_forward + residual_forward?
in which line is this done 😅 ?

Reply:
Oh wait I may have misinterpreted, when you say layer before residual connection, are you talking about F.forward(x)/R.forward(x), or the module before Residual<F, R>?

Reply:
> Oh wait I may have misinterpreted, when you say layer before residual connection, are you talking about F.forward(x)/R.forward(x), or the module before Residual<F, R>?

the module before Residual<F,R> in the forward pass (so the output of that module is the input of Residual<_, _>)

Reply:
> add_grad + f_grad + r_grad

but if the `add(y, &x)` is executed before `2.a` in the backprop, how can f_grad be fetched from gradients to calculate `x`'s gradients for `1`?

Reply:
The three are summed together independently. After add's backprop x's gradient is only 1/3 complete (add_grad). Once f's backward op is executed then x is 2/3 complete (add_grad + f_grad), then finally after r's backward op x is complete (add_grad + f_grad + r_grad).

It works because x is input for add, f, and r, not the result. Gradients only need to be fully "complete" for backward ops in which they are the result.

Reply:
it depends on what f & r are. for binary map the call to addmul does `rhs_grad += rhs_deriv * result_grad`, so if `rhs_grad` is x's grad that's where part of the accumulation happens.

If F and R were both using `map` for instance then both accumulations would happen in [map](https://github.com/coreylowman/dfdx/blob/main/src/tensor_ops/map.rs#L235) (where x's gradient is `g`)

Reply:
> Linear<10, 5>

i mean that linear layer, what gradients will it use? it will be `lhs_grad`, won't it? but when is `rhs_grad` added to `lhs_grad`?